clients/src/test/java/org/apache/kafka/common/config/provider/DirectoryConfigProviderTest.java

```diff
@@ -153,5 +159,57 @@ public void testServiceLoaderDiscovery() {
         ServiceLoader<ConfigProvider> serviceLoader = ServiceLoader.load(ConfigProvider.class);
         assertTrue(StreamSupport.stream(serviceLoader.spliterator(), false).anyMatch(configProvider -> configProvider instanceof DirectoryConfigProvider));
     }
+
+    @Test
+    public void testAllowedPath() {
+        Map<String, String> configs = new HashMap<>();
+        configs.put(ALLOWED_PATHS_CONFIG, parent.getAbsolutePath());
+        provider.configure(configs);
+
+        ConfigData configData = provider.get(dir);
+        assertEquals(toSet(asList(foo, bar)), configData.data().keySet());
+        assertEquals("FOO", configData.data().get(foo));
+        assertEquals("BAR", configData.data().get(bar));
+        assertNull(configData.ttl());
+    }
+
+    @Test
+    public void testMultipleAllowedPaths() {
+        Map<String, String> configs = new HashMap<>();
+        configs.put(ALLOWED_PATHS_CONFIG, dir + "," + siblingDir);
+        provider.configure(configs);
+
+        ConfigData configData = provider.get(subdir);
+        assertEquals(toSet(asList(subdirFileName)), configData.data().keySet());
+        assertEquals("SUBDIRFILE", configData.data().get(subdirFileName));
+        assertNull(configData.ttl());
+
+        configData = provider.get(siblingDir);
+        assertEquals(toSet(asList(siblingDirFileName)), configData.data().keySet());
+        assertEquals("SIBLINGDIRFILE", configData.data().get(siblingDirFileName));
+        assertNull(configData.ttl());
+    }
+
+    @Test
+    public void testNotAllowedPath() {
+        Map<String, String> configs = new HashMap<>();
+        configs.put(ALLOWED_PATHS_CONFIG, dir);
+        provider.configure(configs);
+
+        ConfigData configData = provider.get(siblingDir);
+        assertTrue(configData.data().isEmpty());
+        assertNull(configData.ttl());
+    }
+
+    @Test
+    public void testNoTraversalAllowedPath() {
+        Map<String, String> configs = new HashMap<>();
+        configs.put(ALLOWED_PATHS_CONFIG, dir);
+        provider.configure(configs);
+
+        ConfigData configData = provider.get(dir + Paths.get("/../siblingdir"));
```

Comment:
This is mis-using the Paths.get function, which is intended to do all of the directory delimiting on your behalf.
If the leading `/` was missing in this string, the test would be invalid.

```suggestion
        ConfigData configData = provider.get(Paths.get(dir, "..", "siblingdir").toString());
```

In this form, giving `".."`, `"/.."`, `"../"`, or `"/../"` are all equivalent, so it is more resistant to typos.